src/main/scala/org/apache/spark/sql/kinesis/CachedKinesisProducer.scala

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.kinesis
+
+import java.util.Locale
+import java.util.concurrent.{ExecutionException, TimeUnit}
+
+import scala.collection.JavaConverters._
+import scala.util.control.NonFatal
+
+import com.amazonaws.auth.{AWSStaticCredentialsProvider, BasicAWSCredentials}
+import com.amazonaws.regions.RegionUtils
+import com.amazonaws.services.kinesis.AmazonKinesis
+import com.amazonaws.services.kinesis.producer.{KinesisProducer, KinesisProducerConfiguration}
+import com.google.common.cache._
+import com.google.common.util.concurrent.{ExecutionError, UncheckedExecutionException}
+
+import org.apache.spark.SparkEnv
+import org.apache.spark.internal.Logging
+
+private[kinesis] object CachedKinesisProducer extends Logging {
+
+  private type Producer = KinesisProducer
+
+  private lazy val cacheExpireTimeout: Long =
+    SparkEnv.get.conf.getTimeAsMs("spark.kinesis.producer.cache.timeout", "10m")
```

Comment:
Should we make it Kinesis Sink option?

Also what should be a good value of the timeout. If trigger interval is say 30 secs, because of following code `val writeTask = new KinesisWriteTask(kinesisParameters, schema) 
Utils.tryWithSafeFinally(block = writeTask.execute(iter))(
finallyBlock = writeTask.close())`  cache will be cleared at the end of each batch duration. Do we still need a timeout?

Reply:
The writeTask contains a producer which we get by value from the cache. Once we close the task, the producer inside the cache becomes null, but the producer object in the cache still remains the same until the timeout expires.